### PR TITLE
Schemas

### PR DESCRIPTION
**Pull request by U8NWXD**
We previously had separate wiki pages for schema-based forms and validating handler arguments with schemas. This PR merges those two into a single page documenting how schemas work and how we use them

### CODE REVIEW

**Comment by U8NWXD** [comment]
@vojtechjelinek could you please take a look at my changes to our wiki pages on schemas?

**Review by vojtechjelinek — APPROVED** [review]
LGTM!

File: Schemas.md [hunk 1]
```diff
@@ -0,0 +1,521 @@
...
+      * [Handlers with no arguments](#handlers-with-no-arguments)
+      * [Post schema operations](#post-schema-operations)
+    * [Common errors faced](#common-errors-faced)
+    * [Examples for reference](#examples-for-reference)
+    * [Debugging tricks](#debugging-tricks)
+    * [Contact](#contact)
+
+## Introduction
+
+At Oppia, we often want to describe the possible structure of some data. This data could be as simple as the boolean value of a checkbox in a user's preferences, or it could be more complex like the valid arguments to one of our backend API controllers. We use describe these data using _schemas_.
```
Comment (seanlip) [inline Schemas.md:26]: "use" seems wrong in the last sentence.

File: Schemas.md [hunk 2]
```diff
@@ -0,0 +1,521 @@
...
+
+Schemas describe the structure of data in a machine-readable format so that generic utility functions can operate on arbitrary data so long as the data comes with a schema. For example, schemas let us:
```
Comment (seanlip) [inline Schemas.md:28]: Bit weird to refer to "schemas" at the start and "schema" at the end, in the first sentence. Maybe rephrase so that it flows more cleanly?

File: Schemas.md [hunk 3]
```diff
@@ -0,0 +1,521 @@
...
+
+* Validate provided data to make sure it conforms to the expected schema.
+* Automatically generate forms that let a user provided data in a form specified by a schema.
```
Comment (seanlip) [inline Schemas.md:31]: provided --> provide

File: Schemas.md [hunk 4]
```diff
@@ -0,0 +1,521 @@
...
+    * `dict`
+    * `html`: A string (`bytes` or `str` in Python 3) with HTML code. When normalized, the HTML code will be sanitized by `core.domain.html_cleaner.clean()`. A `bytes` object may also be provided, which will be automatically decoded acording to UTF-8 during normalization.
+    * `custom`: A custom object type defined in `extensions/objects/models/objects.py`.
+    * `object_dict`: A dictionary that corresponds to a domain object (an object whose class is defined in the domain layer).
+    * `unicode_or_none`: Accepts data of type `unicode` or `None`. If the data is not `None`, it is normalized as if it were of type `unicode`.
+
+* Optional keys:
+
+  * `choices`: A list of possible values. The data must exactly match one of the choices.
+  * `validators`: A list of dictionaries, each of which has a sole key `id` that maps to the name of a validation function. The validation functions are defined as static methods of `schema_utils._Validators`. To conform to the schema, data must pass all validators.
```
Comment (seanlip) [inline Schemas.md:56]: ...all the validators specified in the schema.

File: Schemas.md [hunk 5]
```diff
@@ -0,0 +1,521 @@
...
+All arguments passed to the GET/POST/PUT/DELETE methods of the handler classes in the Oppia controller layer or in main.py need to be robustly validated before being passed to the domain layer in the backend. If we define schemas for the expected arguments, then we can use a generic schema validation system (SVS) to validate the arguments!
+
+#### Schema validation system code
+
+The following key methods are used in the validation of handler args through the SVS architecture:
+
+* `vaidate_and_normalize_args()` in `base.py`: This method is defined in the BaseHandler class of base.py. The `vaidate_and_normalize_args` method is responsible for raising all kinds of errors in the context of validation of handler args, like `InvalidInputException` and `NotImplemented` error. (See [this section](#common-errors-faced) for a list of common errors that may arise.)
+
+* `validate(handler_args, handler_args_schemas)` in `payload_validator.py`: This method is the core method for SVS functionality. It collects all the AssertionErrors raised from schema_utils.
+  * `handler_args`: The arguments from payload/ request.
```
Comment (seanlip) [inline Schemas.md:153]: Maybe elaborate on this a bit more, I don't know what `payload/ request` is.

File: Schemas.md [hunk 6]
```diff
@@ -0,0 +1,521 @@
...
+  * `handler_args_schemas`: Schema from the handler class.(See [this link](#how-to-write-validation-schema-for-handlers) for more information on how to write a schema).
+
+* `normalize_against_schema(obj, schema)` in `schema_utils.py`: This method normalizes the obj against its schema and raises `AssertionError` if any of the validation checks fail. This `AssertionError` is represented as `InvalidInputException` to users.
```
Comment (seanlip) [inline Schemas.md:156]: Maybe obj --> given `obj`

File: Schemas.md [hunk 7]
```diff
@@ -0,0 +1,521 @@
...
+
+1. **List all the arguments passed to each method in the handler**
+   Make a list of all the arguments passed to each method in the handler class. Arguments received by a handler class method can be categorized into 3 types:
+
+   * **URL path elements** : The data which is present as a part of the URL are called URL path elements. Example: in ```url/<exploration_id>/```, the exploration_id is a URL path element.
+   * **Payload arguments**: The data which comes from payloads are called payload arguments. These data are typically received by PUT and POST methods.
+   * **URL query parameters**: Query parameters are a defined set of parameters attached to the end of a url. They are extensions of the URL that are used to help define specific content or actions based on the data being passed. Example: in ```url/<exploration_id>?username=nikhil```, there is a single URL query parameter, with arg name “username” and value “nikhil”. URL query parameters are typically received by GET and DELETE methods.
+
+   If you face any difficulty see the [debugging section](#debugging-tricks) or reach out to any of the persons mentioned in the [contact section](#contact).
```
Comment (seanlip) [inline Schemas.md:171]: ...difficulty with this step, see ...

File: Schemas.md [hunk 8]
```diff
@@ -0,0 +1,521 @@
...
+
+2. **Determine the schema for each argument**
+   For writing schema each argument should be analysed deeply, like the use of argument in the backend structure of the code and based on the analysis, schema for the arguments should be written by following the [boilerplate code](#handlers-with-no-arguments). See these links for more information on [allowed schema keys](#write-a-schema), [Important code pointers](#important-code-pointers), and [examples](#examples-for-reference).
```
Comment (seanlip) [inline Schemas.md:174]: I think the grammar/structure of this could be improved.

File: Schemas.md [hunk 9]
```diff
@@ -0,0 +1,521 @@
...
+
+#### Important code pointers
+
+When adding schemas for the args of a particular handler class, some analysis is typically needed. The following points discuss the conventions adopted throughout the codebase for adding schemas to handler classes. **Please read these conventions carefully**.
+
+##### Default and optional arguments
+
+If an argument is not present in a payload/request, and the schema for that argument is defined in the handler, then that argument is treated as “missing”. For missing args, schema utils will raise AssertionError which is represented as InvalidInputException by vaidate_and_normalize_args() method.
+
+To provide default args for a handler, include a key with the name ```default_value``` along with the schema. The value for this key is the default value with which the arg will be updated if no value for that arg is provided in the request. If an argument is optional and it is not supposed to be updated with any default value, then the “default_value” key should contain None.
```
Comment (seanlip) [inline Schemas.md:220]: The first sentence suggests that the key is only provided when we want to provide default args for a handler. But I think we always want to provide that key (based on the examples below)?

File: Schemas.md [hunk 10]
```diff
@@ -0,0 +1,521 @@
...
+}
+```
+
+For more understanding [refer to examples](#examples-for-reference).
+
+##### Domain object arguments: Schemas with type object_dict
+
+Objects which are represented by classes written in the domain layer of the codebase are called domain objects. These classes typically include methods to validate their objects.
+
+For validating domain objects through SVS architecture, there are two preferred solutions, each for their unique cases.
```
Comment (seanlip) [inline Schemas.md:258]: Not sure I understand this sentence.

File: Schemas.md [hunk 11]
```diff
@@ -0,0 +1,521 @@
...
+
+###### Case 1: object_class key
+
+The data coming from the payloads/requests is in the dict format and many of the domain objects do not get initialized with the dictionary form of the data so they must be initialized by using the from_dict() method of the same domain class. In this case class is directly passed into the schema with a schema key named ‘object_class’. Schema for these cases should have the two keys as follows:
```
Comment (seanlip) [inline Schemas.md:262]: First sentence is very run-on, and there seem to be a bunch of missing "the"s. It might be worth doing a grammar audit on this and the rest of this doc?

**Review by seanlip — CHANGES_REQUESTED** [review]
Hi @U8NWXD -- I started a review pass, but then noticed that there are a number of grammar/expression issues in the existing docs. Would you mind taking another pass with an eye towards this, and fixing them?

Thanks!

File: Schemas.md (hunk 9 above)
Comment (U8NWXD) [reply Schemas.md]: I think the first sentence is correct. There are three options for handling missing arguments:

* If there is no `default_value` key, then a missing argument will cause an exception.
* If there is a `default_value` key with the value `None`, then a missing argument will not raise any errors. The normalized data will lack the argument.
* If there is a `default_value` key with a value other than `None` (call the value `x`), then a missing argument will not raise any errors. The normalized data will include the argument with the value `x`.

Here's the relevant code:

https://github.com/oppia/oppia/blob/26b1b5dad4b42a540b2557aba1d27b7bd099f190/core/controllers/payload_validator.py#L58-L67

@vojtechjelinek can you confirm?

File: Schemas.md (hunk 1 above)
Comment (U8NWXD) [reply Schemas.md]: done

File: Schemas.md (hunk 2 above)
Comment (U8NWXD) [reply Schemas.md]: done

File: Schemas.md (hunk 3 above)
Comment (U8NWXD) [reply Schemas.md]: done

File: Schemas.md (hunk 4 above)
Comment (U8NWXD) [reply Schemas.md]: done

File: Schemas.md (hunk 5 above)
Comment (U8NWXD) [reply Schemas.md]: I changed this to `HTTP request` to be clearer

File: Schemas.md (hunk 6 above)
Comment (U8NWXD) [reply Schemas.md]: done

File: Schemas.md (hunk 7 above)
Comment (U8NWXD) [reply Schemas.md]: done

File: Schemas.md (hunk 8 above)
Comment (U8NWXD) [reply Schemas.md]: done

File: Schemas.md (hunk 10 above)
Comment (U8NWXD) [reply Schemas.md]: fixed

File: Schemas.md (hunk 11 above)
Comment (U8NWXD) [reply Schemas.md]: done

**Comment by U8NWXD** [comment]
@seanlip @DubeySandeep I've finished proof-reading the parts I copied from the existing wiki page on schemas for handler args. PTAL

File: Schemas.md [hunk 12]
```diff
@@ -0,0 +1,528 @@
...
+   }
+   ```
+
+#### Important code pointers
+
+The following points discuss the conventions adopted throughout the codebase for adding schemas to handler classes. **Please read these conventions carefully**.
+
+##### Default and optional arguments
+
+If an argument is not present in a request, and the schema for that argument is defined in the handler, then that argument is treated as “missing”. For missing args, `schema_utils` will raise an `AssertionError` which will be represented as an `InvalidInputException` by the `vaidate_and_normalize_args()` method.
```
Comment (seanlip) [inline Schemas.md:220]: Ditto, validate is misspelled

File: Schemas.md [hunk 13]
```diff
@@ -0,0 +1,528 @@
...
+            'schema': {
+                'type': 'bool'
+             },
+            'default_value': False
+        }
+    }
+}
+```
+
+**Example when default value is not provided**: Suppose `make_community_owned` is an optional argument which should not take any default value if no value for that argument is provided with the request. In that case, the schema for `make_community_owned` should look like:
```
Comment (seanlip) [inline Schemas.md:239]: default --> a default

File: Schemas.md [hunk 14]
```diff
@@ -0,0 +1,528 @@
...
+        }
+    }
+}
+```
+
+###### Case 2: `validation_method` key
+
+Sometimes, a domain object has a `validate_dict()` method that should be used for validation. These methods sometimes need extra arguments, for example to enforce strict validation. Since these cases cannot be handled generally, you'll need to write a function that accepts the dictionary from the request and passes it to the domain object's `validate_dict()` method, specifying other parameters as needed. Your function should be in the `domain_objects_validator.py` file.
+
+The newly written function of the `domain_objects_validator.py` file should be directly passed into the schema with a schema key named `validation_method`. Schemas for these cases should have the two keys as follows:
```
Comment (seanlip) [inline Schemas.md:294]: two keys as follows --> two keys set up as follows

(or similar, this might need some thought)

File: Schemas.md [hunk 15]
```diff
@@ -0,0 +1,528 @@
...
+
+1. **type**: 'object_dict'
+2. **validation_method**: method written in domain_objects_validator for calling validate method from domain class directly.
```
Comment (seanlip) [inline Schemas.md:297]: ... the validate() method directly from the domain class.

File: Schemas.md [hunk 16]
```diff
@@ -0,0 +1,528 @@
...
+}
+```
+
+Here, `validate_exploration_change` is the function you wrote.
+
+For more information [refer to our examples](#examples-for-reference).
+
+##### Extra validators
+
+By providing validators, you can increase a schema’s functionality. The `validators` field in the schema contains a list of dicts, where each dict contains a key `id` whose value is the name of the validator. Existing validator methods can be found in the `_Validator` class of  `schema_utils.py`. You can use the existing validators and write new ones.
```
Comment (seanlip) [inline Schemas.md:324]: and write --> or write

File: Schemas.md [hunk 17]
```diff
@@ -0,0 +1,528 @@
...
+
+1. **NotImplementedError**
+
+   * **Description**: This error will be raised if any necessary schemas (i.e, `HANDLER_ARGS_SCHEMAS` or `URL_PATH_ARGS_SCHEMAS`) are not present in the corresponding handler class.
+   * **How to resolve**: This error message is raised with the name of the handler which is missing a schema definition. Add schemas to the specified handler class.
+
+2. **InvalidInputException**
+
+   * **Description**: This error will be raised if schema validation failed for any argument. It may be due to extra arguments, missing arguments or any type mismatch.
+   * **How to resolve**: This error message is raised by the `vaidate_and_normalize_args()` method with the name of the argument for which schema validation failed. By looking at error messages and stack traces, you can find which argument is failing the schema validation test and debug.
```
Comment (seanlip) [inline Schemas.md:384]: validate is misspelled. (Maybe fix throughout?)

File: Schemas.md [hunk 18]
```diff
@@ -0,0 +1,528 @@
...
+
+#### Examples for reference
+
+Examples of PRs for different types is given below:
+
+* [Sample pr 1](https://github.com/oppia/oppia/pull/13223)
```
Comment (seanlip) [inline Schemas.md:390]: pr --> PR (here and below)

File: Schemas.md [hunk 19]
```diff
@@ -0,0 +1,528 @@
...
+   ('payload', u'{"version":1,"new_member_role":"owner","new_member_username":"nikhil"}')
+   ('source', u'http://localhost:8181/create/QuWbhgRTovXr#/settings')
+   (u'URL path elements = ', {u'exploration_id': 'QuWbhgRTovXr'})
+   (u'Request method = ', 'PUT')
+   (u'HANDLER_ARGS_SCHEMAS =  ', {u'DELETE': {u'username': {u'type': u'basestring'}}, u'PUT': {u'make_community_owned': {u'default_value': None, u'type': u'bool'}, u'new_member_role': {u'default_value': None, u'type': u'basestring'}, u'new_member_username': {u'default_value': None, u'type': u'basestring'}, u'viewable_if_private': {u'default_value': None, u'type': u'bool'}, u'version': {u'type': u'int'}}})
+   (u'URL_PATH_ARGS_SCHEMAS = , ', {u'exploration_id': {u'type': u'basestring'}})
+   ------------------------------------
+   ```
+
+6. **Write schema by following the boilerplate code**
```
Comment (seanlip) [inline Schemas.md:451]: schema --> the schema

File: Schemas.md [hunk 20]
```diff
@@ -0,0 +1,528 @@
...
+   Writing the schema is the most crucial part, and it is important to get this correct. The print logs from the previous step can help you get started, but please be sure to dig into the backend and frontend code, and follow calls to methods/functions to see how the incoming data is used. This will help you avoid making errors. In particular:
+
+   * **For the backend**: Try to read code as well as docstrings of all the methods which use the arguments from request.
```
Comment (seanlip) [inline Schemas.md:454]: request --> the request

File: Schemas.md [hunk 21]
```diff
@@ -0,0 +1,528 @@
...
+
+7. Remove print statements
+
+Remove all the print statements and verify schema validation by again hitting the handler from the frontend.
+
+#### Contact
+
+For any discussion please contact one of:
+
+* Rohit(@rohitkatlaa)
```
Comment (seanlip) [inline Schemas.md:522]: Add spaces after each name and before each '('?

File: Schemas.md [hunk 22]
```diff
@@ -0,0 +1,528 @@
...
+* Vojtech(@vojtechjelinek)
+* Nikhil(@Nik-09).
+
+### Configuration values
```
Comment (seanlip) [inline Schemas.md:526]: This seems weirdly positioned. Why is it all the way down here?

**Review by seanlip — CHANGES_REQUESTED** [review]
Thanks! Looks better, I took a full pass.

**Comment by seanlip** [comment]
@vojtechjelinek also assigning you for https://github.com/U8NWXD/oppia-wiki/pull/16#discussion_r693221668

File: Schemas.md (hunk 9 above)
Comment (vojtechjelinek) [reply Schemas.md]: @Nik-09 Can you confirm Chrises assumption?

Comment (Nik-09) [reply Schemas.md]: True, the three points mentioned by @U8NWXD aligns with the idea for writing schema for handler params.
Thank you.
